### PR TITLE
Alt clicking on doors sends the AI a request link to open it

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
:cl: Kor
rscadd: Alt+clicking on an airlock will now send a request to the AI to open it. The door must be on camera for this to work.
/:cl:

Alt+click on a door will send the AI a notice with the doors name, and the name (not real name, so disguises will work here) of the person requesting the door. Clicking the link will open the door.

There is a 10 second cooldown on each door to stop request spam.

The door must be near cameras, or the AI can not open it.

Reasoning: AI DOOR followed by using your camera to track them is awful with the artificial lag we added to finding them, but I don't want to buff the AI by removing that entirely when its responding to HUNT DOWN THIS CRIMINAL.

I dont even like the AI very much but AI DOOR is not a very interesting or important part of gameplay, it should be streamlined.

### CODE REVIEW

**Comment by TheNightingale** [comment]
So someone who doesn't have a radio or is near an intercom... can get the AI to open a door for them? Not sure we need to hold peoples' hands through opening doors.

**Comment by ChangelingRain** [comment]
AI DOOR AI, AI, DOOR

**Comment by Dorsisdwarf** [comment]
:-1: 
Use a radio you shitters

**Comment by KazeEspada** [comment]
Who needs a radio when we have buttons!

**Comment by Buggy123** [comment]
:+1: But add a option to open the door controls so AIs can cuck people by shocking them instead of opening them.

**Comment by AnturK** [comment]
Should probably display requesting person job/id too. I guess it doesn't matter for asimov but will save headaches.

**Comment by TheNightingale** [comment]
When you say "10 second cooldown on each door".... does that mean one person can run between lots of different doors and alt-click them all?

**Comment by optimumtact** [comment]
That's an ingenious idea.

**Comment by PKPenguin321** [comment]
:-1: You can't take people's ID and headset and lock them into rooms anymore

**Comment by KorPhaeron** [comment]
You can if you knock out the camera

**Comment by KorPhaeron** [comment]
@TheNightingale

Yes, but if someone abuses this to spam, admins can deal with it the same as any other spam in the game

**Comment by TheNightingale** [comment]
People already abuse 'showing paper to a camera' to spam. This seems too easy; if someone wants a door open, why can't they just ask for it ";AI DOAH"?

It also makes cutting comms less effective, if anyone can just alt-click a nearby door to get the AI eye on them (e.g. while they're being killed).

**Comment by KorPhaeron** [comment]
>why can't they just ask for it ";AI DOAH"?

I said why in the OP

>people already abuse

And we ban them

>if anyone can alt click a door

You can't be stunned/unconscious/etc, and I believe canusetopic also requires you to be adjacent, but I will double check.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -418,6 +418,17 @@ var/list/ai_list = list()
 				H.attack_ai(src) //may as well recycle
 			else
 				src << "<span class='notice'>Unable to locate the holopad.</span>"
+		return
+
+	if(href_list["remotedoor"])
+		var/obj/machinery/door/airlock/A = locate(href_list["remotedoor"])
```
Comment (Chnkr) [inline code/modules/mob/living/silicon/ai/ai.dm]: Yes, lets allow the AI to open **all** doors with a few topic calls, what could go wrong?

Comment (GunHog) [inline code/modules/mob/living/silicon/ai/ai.dm]: There is a Malf power that does this, sort of.

**Comment by GunHog** [comment]
**AI, DOOR! AI, DOOR!! AI!! DOOOOOOORRR!!!** :+1:

File: code/modules/mob/living/silicon/ai/ai.dm (hunk 1 above)
Comment (KorPhaeron) [inline code/modules/mob/living/silicon/ai/ai.dm]: @neersighted 

Said he was going to show me how to fix potential href exploits

**Comment by xxalpha** [comment]
Make it require an ID to send a request and use the information from the ID (Name and Job Name) to notify the AI, and add a cooldown per door and per ID.

:+1:

**Comment by KorPhaeron** [comment]
@xxalpha 

What if I want Ian to be able to request doors?

**Comment by duncathan** [comment]
Make it send a radio message so that if comms are cut people can't access their doorknob

**Comment by xxalpha** [comment]
Fuck ian.

File: code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -1128,4 +1130,29 @@ var/list/airlock_overlays = list()
...
+/obj/machinery/door/airlock/AltClick(mob/living/user)
+	if(!istype(user))
+		user << "<span class='info'>Nice try, ghosts.</span>"
+		return
+
+	if (!user.canUseTopic(src))
+		user << "<span class='info'>You can't do this right now!</span>"
+		return
+
+	if(stat & (NOPOWER|BROKEN) || emagged)
```
Comment (GunHog) [inline code/game/machinery/doors/airlock.dm]: Check that the AI wire is not cut.

**Comment by Cebasiz** [comment]
i play sillicons and this doesn't seem like a bad idea to me. 

in highpop door requests often get dropped because of the camera track lag, especially if everyone happens to be on the other side of the station from the last guy.

which happens more than you might think.

**Comment by freerealestate** [comment]
Great work :+1:

**Comment by MrPerson** [comment]
You know, if it's so assumed that the AI will open a door that all it needs to see is the name of the door, I've gotta wonder if maybe that door should have just been public to begin with.

**Comment by xxalpha** [comment]
By default it has to open the doors, but that is not always the case.

**Comment by AlexanderUlanH** [comment]
:-1: "AI DOOR" requests have always been partly balanced around the fact that everyone can hear that you're trying to get into somewhere.  It also gives people who should be overlooked on account of having no radio an easy way to get the AI's attention.  Expect a lot of this being used from perma, and in the few occasions where an antag decides to leave someone alive but stranded.
Plus, it looks like you can even do this to doors not on camera network.  Who cares that the AI can't open it? It'll send a borg or a human, and the traitor who reasonably figured this was a fine way to get rid of someone for a while will just wind up screwed.

I'd suggest we instead remove the delay from follow links but keep it for trying to follow with a name, so that the AI can jump to "AI door" instantly but still has to wait a few seconds to respond to "AI TRACK TRAITOR"

Also "we can ban them if they do" is not a reason to leave something abusable in the code.  If we're doing this, why not just make it a ten second cooldown per person?

**Comment by KorPhaeron** [comment]
>Also "we can ban them if they do" is not a reason to leave something abusable in the code.

We would have to just scrap the game entirely if we tried to remove everything abuseable.

>looks like you can do it to doors away from cameras

At least read the damn PR before you complain about it

**Comment by Jazaen** [comment]
Maybe make bolted doors un-requestable, so traitor/subverted/irritated AI can make lizard perma prisoners shut up.

**Comment by swankcookie** [comment]
:+1: this is ace

**Comment by borntohonk** [comment]
make it ignore requests from lizards

**Comment by Shadowlight213** [comment]
@KorPhaeron

**Comment by KorPhaeron** [comment]
Stale PR closing it unless the lazy OP says he wants it re-opened

**Comment by KazeEspada** [comment]
NNNNNNNNNNOOOOOO

**Comment by PKPenguin321** [comment]
:ok_hand:

**Pull request closed without merging**